keps/sig-apps/4368-support-managed-by-label-for-batch-jobs/README.md

```diff
@@ -0,0 +1,985 @@
+feature with similar characteristics.
+
+Finally, we would like to have the feature soak without unnecessary delay.
+
+#### Can the label be mutable?
+
+There is a potential risk of leaking pods, if the value of the label is changed.
+For example, assume there is a running Job, which is reconciled by the Job
+controller, and has some pods created.
+Then, if the label is switched to the mirroring Kueue controller (which by
+itself does not manage pods). Then, the pods are leaking and remain running.
+
+In order to avoid the risk of pods leaking between the controllers when changing
+value of the `managed-by` label, we make the label immutable (see also
+[label mutability](#label-mutability)).
+
+However, the question remains if we can make the label mutable when the job is
+stopped, similarly, as we do with the `AllowMutableSchedulingDirectives` flag
+which guards mutability of the Job's pod template labels.
+
+It seems possible, and could be handy in [Story 1](#story-1), but it is also not
+a blocker.
+
+It would also the complicate debuggability of the feature.
+
+We decide to keep the label immutable, at least for [Beta](#beta), we will
+re-evaluate the decision for [GA](#ga).
+
+### Risks and Mitigations
+
+#### Ecosystem fragmentation due to forks
+
+The mechanism to disable the main Job controller opens the door for users to
+substitute it with a fork. This may create more fragmentation in the community
+as users may prefer to use their forked controllers rather than contribute
+upstream.
+
+First, this risk, to some extent, exists even today as admins with access to the
+control plane can disable by passing `--controllers=-job,*` in the manifest for
+`kube-controller-manager` (see more info on the `--controllers` flag
+[here](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/)).
+
+Second, we believe that users who had the need to fork the Job controller
+already introduced dedicated Job CRDs for their needs.
+
+#### Two controllers running at the same time on old version
+
+It is possible that one configures MultiKueue (or another project) with an
+older version of k8s which does not support the label yet. In that case
+two controllers might start running and compete with Job status updates at the
+same time.
+
+To mitigate this risk we warn about it in Kueue documentation, to use this label
+only against newer versions of Kubernetes.
+
+Finally, this risk will fade away with time as the new versions of
+Kubernetes support it.
+
+## Design Details
+
+#### API
+
+```golang
+const (
+  ...
+	// LabelManagedBy is used to indicate the controller or entity that manages
+	// an Job.
+	LabelManagedBy = "batch.kubernetes.io/managed-by"
+)
+```
+
+#### Implementation overview
+
+We skip synchronization of the Jobs with the `managed-by` label, if it has any
+different value than `job-controller.k8s.io`. When the synchronization is skipped,
+the name of the controller managing the Job object is logged.
```

Comment:
> Debugging is important , for this purpose we will log the line saying that the syncing is delegated to another controller (available to admins / OnCalls).

Logs might be overlooked. Wouldn't emitting an event be better for the debugging?

If we have the event immutable once the object is created, we can fire the event on creation only. Or we could fire it on change of the label.

Reply:
Emitting events seems like an idea worth considering. However they expire, and I imagine in most cases users would want to inspect their jobs after they were stuck for a significant amount of time. Renewing the event on every update might be adding too much of performance penalty. So, when playbooks are written to investigate if a job is stuck, then events wouldn't be a very reliable source of information. 

One good thing about using events is that they get displayed in `kubectl` describe. I think we can achieve this by modifying the `kubectl` describe to add a human readable information on the bottom, based on the label value.

Reply:
I agree, firing the event repeatedly on syncs would not be very beneficial.  Even though the events are transient, single event emission might still help to debug the problem in some cases. Also, there is a precedent for firing events in similar scenarios in our controllers.

There is also another drawback for users who would use this feature often, they might find the events spammy.

Reply:
> Also, there is a precedent for firing events in similar scenarios in our controllers.

could you provide an example? One situation similar to what we do is custom `spec.schedulerName`, but I tested and we don't seem to fire events if the value is non-default.

>  they might find the events spammy.

I think is a valid concern for repeatedly fired event on every sync.

Reply:
> could you provide an example? One situation similar to what we do is custom spec.schedulerName, but I tested and we don't seem to fire events if the value is non-default.

I did not mean for this specific use case, but that we use events in general to communicate abnormal cases/behaviour. Like here:

https://github.com/kubernetes/kubernetes/blob/fbb2e6293fb0c8c107ae48b8b8ae488325c59598/pkg/controller/disruption/disruption.go#L587-L589

Reply:
I see, but the situation you reference is a warning, so seems justified to be spammy to let the user know. 

In this case of `managed-by` it would be spammy in the happy path.